File: backend/density.go
package main

import (
	"math"
)

const (
	densityMean = 0.20
	sScale      = 0.20

	// Length scales in chunk units
	broadLenChunks = 20
	localLenChunks = 2.5

	// Weights
	aWeight = 0.6
	bWeight = 0.8
)

// Quintic fade 6t^5 - 15t^4 + 10t^3
func fade(t float64) float64 { return t * t * t * (t*(t*6-15) + 10) }

// gradAngle creates a deterministic gradient angle in [0,2π) from lattice (xi,yi) and seed
func gradAngle(xi, yi int64, seed uint64) float64 {
	const A = uint64(0x1F123BB5)
	const B = uint64(0x9E3779B1)
	const C = uint64(0x94D049BB)
	h := splitmix64(uint64(uint64(xi)*A + uint64(yi)*B + seed*C))
	const inv2p64 = 1.0 / 18446744073709551616.0 // 2^64
	return (float64(h) * inv2p64) * (2.0 * math.Pi)
}

func gradDot(xi, yi int64, x, y float64, seed uint64) float64 {
	a := gradAngle(xi, yi, seed)
	gx, gy := math.Cos(a), math.Sin(a)
	return gx*(x-float64(xi)) + gy*(y-float64(yi))
}

// gradientNoise2D: Perlin-style single-octave gradient noise with quintic interpolation
// Approximately zero-mean with range around [-1,1]
func gradientNoise2D(x, y, lengthScale float64, seed uint64) float64 {
	s := lengthScale
	xf := x / s
	yf := y / s

	x0 := math.Floor(xf)
	y0 := math.Floor(yf)
	x1 := x0 + 1
	y1 := y0 + 1
	tx := xf - x0
	ty := yf - y0
	u := fade(tx)
	v := fade(ty)

	n00 := gradDot(int64(x0), int64(y0), xf, yf, seed)
	n10 := gradDot(int64(x1), int64(y0), xf, yf, seed)
	n01 := gradDot(int64(x0), int64(y1), xf, yf, seed)
	n11 := gradDot(int64(x1), int64(y1), xf, yf, seed)

	nx0 := n00 + u*(n10-n00)
	nx1 := n01 + u*(n11-n01)
	return nx0 + v*(nx1-nx0)
}

// getChunkDensity returns the canonical per-chunk bomb density in [0,1]. Deterministic and cached.
func (s *Server) getChunkDensity(chunkID ChunkID) float64 {
	s.densityCacheMu.RLock()
	if d, ok := s.densityCache[chunkID]; ok {
		s.densityCacheMu.RUnlock()
		return d
	}
	s.densityCacheMu.RUnlock()

	cx := float64(chunkID.X)
	cy := float64(chunkID.Y)
	broad := gradientNoise2D(cx, cy, broadLenChunks, 42)
	local := gradientNoise2D(cx, cy, localLenChunks, 42^0xA5A5A5A5)

	// Combine independent unit-variance fields, then normalize to unit std
	combined := aWeight*broad + bWeight*local
	norm := math.Sqrt(aWeight*aWeight + bWeight*bWeight)
	if norm > 0 {
		combined /= norm
	}

	dens := densityMean + sScale*combined
	// Clamp to [0,1] for safety
	if dens < 0 {
		dens = 0
	} else if dens > 1 {
		dens = 1
	}

	s.densityCacheMu.Lock()
	s.densityCache[chunkID] = dens
	s.densityCacheMu.Unlock()
	return dens
}

// getScoreMultiplier returns a small, controllable multiplier based on density (centered at 1.0).
func (s *Server) getScoreMultiplier(chunkID ChunkID) float64 {
	d := s.getChunkDensity(chunkID)
	// One-sigma above mean (~+0.0194) gives +k multiplier
	const k = 0.20 // conservative for now; tune via config if needed
	m := 1.0 + k*((d-densityMean)/sScale)
	if m < 0.75 {
		m = 0.75
	} else if m > 1.5 {
		m = 1.5
	}
	return m
}

File: scripts/python/density.py
import math

import matplotlib.pyplot as plt
import numpy as np

W = H = 20
MEAN_DENSITY = 0.20
STD_DENSITY = 0.20
BROAD_LEN = 20  # larger = smoother
LOCAL_LEN = 2.5  # smaller = more local variation
BROAD_WEIGHT = 0.6
LOCAL_WEIGHT = 0.8

_SEED = np.random.randint(0, 2**32)  # 42
_SEED2 = _SEED ^ 0xA5A5A5A5


def splitmix64(n: int) -> int:
    n = (n + 0x9E3779B97F4A7C15) & 0xFFFFFFFFFFFFFFFF
    n ^= n >> 30
    n = (n * 0xBF58476D1CE4E5B9) & 0xFFFFFFFFFFFFFFFF
    n ^= n >> 27
    n = (n * 0x94D049BB133111EB) & 0xFFFFFFFFFFFFFFFF
    n ^= n >> 31
    return n & 0xFFFFFFFFFFFFFFFF


def grad_angle(xi: int, yi: int, seed: int) -> float:
    # Map hashed lattice point → angle in [0, 2π)
    h = splitmix64(
        (xi * 0x1F123BB5 + yi * 0x9E3779B1 + seed * 0x94D049BB) & 0xFFFFFFFFFFFFFFFF
    )
    return (h / 2**64) * (2.0 * math.pi)


def fade(t: float) -> float:
    # Quintic smoothstep 6t^5 - 15t^4 + 10t^3
    return t * t * t * (t * (t * 6 - 15) + 10)


def grad_dot(xi: int, yi: int, x: float, y: float, seed: int) -> float:
    a = grad_angle(xi, yi, seed)
    gx, gy = math.cos(a), math.sin(a)
    return gx * (x - xi) + gy * (y - yi)


def gradient_noise_2d(x: float, y: float, length_scale: float, seed: int) -> float:
    """
    2D gradient (Perlin-style) noise:
      - gradients from stateless hash
      - quintic fade interpolation
      - approx output in [-1,1], mean ~0
    """
    s = length_scale
    xf = x / s
    yf = y / s

    x0 = math.floor(xf)
    y0 = math.floor(yf)
    x1 = x0 + 1
    y1 = y0 + 1
    tx = xf - x0
    ty = yf - y0
    u = fade(tx)
    v = fade(ty)

    n00 = grad_dot(x0, y0, xf, yf, seed)
    n10 = grad_dot(x1, y0, xf, yf, seed)
    n01 = grad_dot(x0, y1, xf, yf, seed)
    n11 = grad_dot(x1, y1, xf, yf, seed)

    nx0 = n00 + u * (n10 - n00)
    nx1 = n01 + u * (n11 - n01)
    return nx0 + v * (nx1 - nx0)


# Keep combined amplitude stable
_norm = math.sqrt(BROAD_WEIGHT**2 + LOCAL_WEIGHT**2) + 1e-12


def get_mine_density(x: int, y: int) -> float:
    broad = gradient_noise_2d(float(x), float(y), BROAD_LEN, _SEED)
    local = gradient_noise_2d(float(x), float(y), LOCAL_LEN, _SEED2)
    z = (BROAD_WEIGHT * broad + LOCAL_WEIGHT * local) / _norm  # ~zero-mean, unit-ish std
    d = MEAN_DENSITY + STD_DENSITY * z
    # Light clamp for safety
    return 0.0 if d < 0.0 else (1.0 if d > 1.0 else d)


densities = np.zeros((H, W), dtype=np.float32)
for y in range(H):
    for x in range(W):
        cx = x - W // 2
        cy = y - H // 2
        densities[y, x] = get_mine_density(cx, cy)

mean_pct = densities.mean() * 100
std_pct = densities.std() * 100

dx = densities[:, 1:] - densities[:, :-1]
dy = densities[1:, :] - densities[:-1, :]
adj_std = np.concatenate([dx.ravel(), dy.ravel()]).std() * 100

print(f"Mean: {mean_pct:.2f}%  Std: {std_pct:.2f}%")
print(f"Adjacent diff std: {adj_std:.3f}pp")

fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(8, 10))
im = ax1.imshow(
    densities * 100, origin="lower", extent=[-W // 2, W // 2 - 1, -H // 2, H // 2 - 1]
)
ax1.set_title(f"Mine Density ({W}×{H}), Centered at (0,0)")
ax1.set_xlabel("x (centered)")
ax1.set_ylabel("y (centered)")
plt.colorbar(im, ax=ax1, label="Density (%)")

ax2.hist((densities * 100).ravel(), bins=40)
ax2.set_title("Density Distribution")
ax2.set_xlabel("Density (%)")
plt.tight_layout()
plt.savefig("density_map.png")
